File: fstar/_output/ocaml/Space_Discipline.ml
open Prims
let can_copy (d : Space_Types.discipline) : Prims.bool=
  match d with
  | Space_Types.Unrestricted -> true
  | Space_Types.Affine -> false
  | Space_Types.Linear -> false
let can_drop (d : Space_Types.discipline) : Prims.bool=
  match d with
  | Space_Types.Unrestricted -> true
  | Space_Types.Affine -> true
  | Space_Types.Linear -> false
let requires_release (d : Space_Types.discipline) : Prims.bool=
  match d with
  | Space_Types.Unrestricted -> false
  | Space_Types.Affine -> true
  | Space_Types.Linear -> false
let self_destructs (d : Space_Types.discipline) : Prims.bool=
  match d with
  | Space_Types.Unrestricted -> false
  | Space_Types.Affine -> false
  | Space_Types.Linear -> true
let more_restrictive (d1 : Space_Types.discipline)
  (d2 : Space_Types.discipline) : Prims.bool=
  match (d1, d2) with
  | (Space_Types.Linear, uu___) -> true
  | (Space_Types.Affine, Space_Types.Unrestricted) -> true
  | (uu___, uu___1) -> false

File: src/_output/fsharp/Space_Discipline.fs
#light "off"
module Space_Discipline

let can_copy : Space_Types.discipline  ->  Prims.bool = (fun ( d  :  Space_Types.discipline ) -> (match (d) with
| Space_Types.Unrestricted -> begin
true
end
| Space_Types.Affine -> begin
false
end
| Space_Types.Linear -> begin
false
end))


let can_drop : Space_Types.discipline  ->  Prims.bool = (fun ( d  :  Space_Types.discipline ) -> (match (d) with
| Space_Types.Unrestricted -> begin
true
end
| Space_Types.Affine -> begin
true
end
| Space_Types.Linear -> begin
false
end))


let requires_release : Space_Types.discipline  ->  Prims.bool = (fun ( d  :  Space_Types.discipline ) -> (match (d) with
| Space_Types.Unrestricted -> begin
false
end
| Space_Types.Affine -> begin
true
end
| Space_Types.Linear -> begin
false
end))


let self_destructs : Space_Types.discipline  ->  Prims.bool = (fun ( d  :  Space_Types.discipline ) -> (match (d) with
| Space_Types.Unrestricted -> begin
false
end
| Space_Types.Affine -> begin
false
end
| Space_Types.Linear -> begin
true
end))


let more_restrictive : Space_Types.discipline  ->  Space_Types.discipline  ->  Prims.bool = (fun ( d1  :  Space_Types.discipline ) ( d2  :  Space_Types.discipline ) -> (match (((d1), (d2))) with
| (Space_Types.Linear, uu___) -> begin
true
end
| (Space_Types.Affine, Space_Types.Unrestricted) -> begin
true
end
| (uu___, uu___1) -> begin
false
end))




